akd_audit_quorum/src/crypto/mod.rs

```diff
@@ -0,0 +1,309 @@
+#[async_trait]
+pub trait QuorumCryptographer: Send + Sync + Clone {
+    // ==================================================================
+    // To be implemented
+    // ==================================================================
+
+    /// Retrieve the public key of this quorum node
+    async fn retrieve_public_key(&self) -> Result<Vec<u8>, QuorumOperationError>;
+
+    /// Retrieve the public key of the Quorum Key
+    async fn retrieve_qk_public_key(&self) -> Result<Vec<u8>, QuorumOperationError>;
+
+    /// Retrieve this node's shard of the quorum key from persistent secure storage
+    async fn retrieve_qk_shard(
+        &self,
+        node_id: NodeId,
+    ) -> Result<EncryptedQuorumKeyShard, QuorumOperationError>;
+
+    /// Save this node's shard of the quorum key to persistent secure storage
+    async fn update_qk_shard(
+        &self,
+        shard: EncryptedQuorumKeyShard,
+    ) -> Result<(), QuorumOperationError>;
+
+    /// Generate the encrypted shards for nodes [0, node_public_keys.len()) encrypted with the public keys of the nodes.
+    /// The provided shards are the quorum key shards, encrypted with THIS node's public key as the leader.
+    async fn generate_encrypted_shards(
+        &self,
+        shards: Vec<EncryptedQuorumKeyShard>,
+        node_public_keys: Vec<Vec<u8>>,
+    ) -> Result<Vec<EncryptedQuorumKeyShard>, QuorumOperationError>;
+
+    /// Encrypt the given material using the provided public key, optionally with the provided nonce
+    async fn encrypt_message(
+        &self,
+        public_key: Vec<u8>,
+        plaintext: Vec<u8>,
+        nonce: Nonce,
+    ) -> Result<Vec<u8>, QuorumOperationError>;
+
+    /// Decrypt the specified message utilizing this node's
+    /// private key. This will require that the message contain
+    /// a nonce. I.e. this cannot be used to decrypt the encrypted
+    /// shard partials
+    async fn decrypt_message(
+        &self,
+        cyphertext: Vec<u8>,
+    ) -> Result<(Vec<u8>, Nonce), QuorumOperationError>;
+
+    /// Generate a commitment on the epoch changes using the quorum key
+    async fn generate_commitment<H: Hasher>(
+        &self,
+        quorum_key_shards: Vec<EncryptedQuorumKeyShard>,
+        epoch: u64,
+        previous_hash: H::Digest,
+        current_hash: H::Digest,
+    ) -> Result<QuorumCommitment<H>, QuorumOperationError>;
+
+    /// Validate the commitment applied on the specified epoch settings
+    async fn validate_commitment<H: Hasher>(
+        public_key: Vec<u8>,
+        commitment: QuorumCommitment<H>,
+    ) -> Result<bool, QuorumOperationError>;
+
+    // ==================================================================
+    // Common trait logic
+    // ==================================================================
+
+    /// Get the number of shards required to reconstruct the quorum key
+    fn shards_required(n: u8) -> u8 {
+        let f = if (n - 1) % 3 == 0 {
```

Comment:
let's add some input validation, like n >= 3 ?

Reply:
It currently returns 0 for n = 1 - is this expected? Do we even need SSS for that case?
I'd expect SSS to be applicable for 2 machines at min.